engine/execution/ingestion/engine.go

```diff
@@ -180,6 +186,130 @@ func (e *Engine) process(originID flow.Identifier, event interface{}) error {
 	}
 }
 
+// on nodes startup, we need to load all the unexecuted blocks to the execution queues.
+// blocks have to be loaded in the way that the parent has been loaded before loading its children
+func (e *Engine) reloadUnexecutedBlocks() error {
+	err := e.reloadFinalizedUnexecutedBlocks()
+	if err != nil {
+		return fmt.Errorf("could not reload finalized unexecuted blocks")
+	}
+
+	err = e.reloadPendingUnexecutedBlocks()
+	if err != nil {
+		return fmt.Errorf("could not reload pending unexecuted blocks")
+	}
+
+	return nil
+}
+
+func (e *Engine) reloadFinalizedUnexecutedBlocks() error {
+	// get finalized height
+	final, err := e.state.Final().Head()
+	if err != nil {
+		return fmt.Errorf("could not get finalized block: %w", err)
+	}
+
+	// find the first unexecuted and finalized block
+	// we iterate from the last finalized, check if it has been executed,
+	// if not, keep going to the lower height, until we find an executed
+	// block, and then the next height is the first unexecuted.
+	// if there is only one finalized, and it's executed (i.e. genesis),
+	// then the firstUnexecuted is a unfinalized block, which is ok,
+	// because the next loop will ensure it only iterate through finalized
+	// block.
+	firstUnexecuted := final.Height
+	for ; ; firstUnexecuted-- {
```

Comment:
Can this loop turn into infinite one? I mean, if no block is found (ie. empty database or some other kind of error) - we will never break out of it.
Its unlikely, but still can happen - shall we add some form of protection, even if its just checking we don't go below zero or so?

Reply:
yeah, i thought about this as well. And I don't think it will be infinite loop.

Normally, the root block is always executed already, so if we reach the root block, it will exit the loop

And the function body also requires the block header at that height to exist. So if it went underflow or no block is found, which both are exception cases, then the query to find block header will return error and then exit the loop. 

But I agree it's probably safer to ensure it's never go below zero.

Reply:
```
engine/execution/ingestion/engine.go:205:8: SA4003: every value of type uint64 is >= 0 (staticcheck)
        for ; lastExecuted >= 0; lastExecuted-- {
              ^
```
Actually, would cause this error